frame/merkle-mountain-range/src/mmr/storage.rs

```diff
@@ -58,16 +72,146 @@ impl<StorageType, T, I, L> Default for Storage<StorageType, T, I, L> {
 	}
 }
 
+impl<T, I, L> Storage<OffchainStorage, T, I, L>
+where
+	T: Config<I>,
+	I: 'static,
+	L: primitives::FullLeaf,
+{
+	/// Move nodes and leaves added by block `N` in offchain db from _fork-aware key_ to
+	/// _canonical key_,
+	/// where `N` is `frame_system::BlockHashCount` blocks behind current block number.
+	///
+	/// This "canonicalization" process is required because the _fork-aware key_ value depends
+	/// on `frame_system::block_hash(block_num)` map which only holds the last
+	/// `frame_system::BlockHashCount` blocks.
+	///
+	/// Should only be called from offchain context, because it requires both read and write
+	/// access to offchain db.
+	pub(crate) fn canonicalize_offchain(block: T::BlockNumber) {
+		use sp_runtime::traits::UniqueSaturatedInto;
+
+		let map_key = Pallet::<T, I>::pruning_map_offchain_key();
+		let mut pruning_map: PruningMap<T> =
+			offchain::local_storage_get(StorageKind::PERSISTENT, &map_key)
+				.and_then(|v| codec::Decode::decode(&mut &*v).ok())
+				.unwrap_or_default();
+
+		// Add "block_num -> hash" mapping to offchain db,
+		// with all forks pushing hashes to same entry (for a particular block number).
+		let parent_hash = <frame_system::Pallet<T>>::parent_hash();
+		pruning_map
+			.entry(block)
+			.and_modify(|e| e.push(parent_hash))
+			.or_insert(vec![parent_hash]);
+
+		// Effectively move a rolling window of fork-unique leaves. Once out of the window, leaves
+		// are "canonicalized" in offchain by moving them under `Pallet::node_canon_offchain_key`.
+		let leaves = NumberOfLeaves::<T, I>::get();
+		let window_size =
+			<T as frame_system::Config>::BlockHashCount::get().unique_saturated_into();
+		if leaves >= window_size {
+			// Move the rolling window towards the end of  `block_num->hash` mappings available
+			// in the runtime: we "canonicalize" the leaf at the end,
+			let to_canon_leaf = leaves.saturating_sub(window_size);
+			// and all the nodes added by that leaf.
+			let to_canon_nodes = NodesUtils::right_branch_ending_in_leaf(to_canon_leaf);
+			frame_support::log::debug!(
+				target: "runtime::mmr", "Nodes to canon for leaf {}: {:?}",
+				to_canon_leaf, to_canon_nodes
+			);
+			// For this block number there may be node entries saved from multiple forks.
+			let to_canon_block_num =
+				Pallet::<T, I>::leaf_index_to_parent_block_num(to_canon_leaf, leaves);
+			// Only entries under this hash (retrieved from state on current canon fork) are to be
+			// persisted. All other entries added by same block number will be cleared.
+			let to_canon_hash = <frame_system::Pallet<T>>::block_hash(to_canon_block_num);
+
+			Self::canonicalize_nodes_for_hash(&to_canon_nodes, to_canon_hash);
+			// Get all the forks to prune, also remove them from the pruning_map.
+			pruning_map
+				.remove(&to_canon_block_num)
+				.map(|forks| {
+					Self::prune_nodes_for_forks(&to_canon_nodes, forks);
```

Comment:
`canonicalize_nodes_for_hash()` doesn't clear/prune anymore, it only populates canon entry, leaving _all_ pruning to be done here.

Right now:
```rust
fn canonicalize_nodes_for_hash(bla) {
    for nodes in bla {
        val = get(fork_x_node_key);
        set(canon_node_key, val)
    }
}

fn prune_nodes_for_forks(fork) { // is called for all forks including 'fork_x' canonicalized above
    for nodes in bla {
        clear(fork_node_key);
    }
}
```

Do you think it's more efficient to do:
```rust
fn canonicalize_nodes_for_hash(bla) {
    for nodes in bla {
        val = get(fork_x_node_key);
        clear(fork_x_node_key);
        set(canon_node_key, val)
    }
}

fn prune_nodes_for_forks(fork) { // is NOT called for 'fork_x_key' canonicalized above
    for nodes in bla {
        clear(fork_node_key);
    }
}
```
?

Reply:
I would think the second one is more efficient (db access are more local to each other), but first one may be better, because code looks cleaner and the number of node is so small it will probably/certainly makes no difference.